delfin/drivers/huawei/oceanstor/oceanstor.py

```diff
@@ -178,6 +178,209 @@ def list_volumes(self, context):
+
+        except Exception as err:
+            LOG.error("Failed to get controller metrics from OceanStor: {}"
+                      .format(err))
+            raise exception.StorageBackendException(
+                reason='Failed to get controller metrics from OceanStor')
+
+    def list_ports(self, context):
+        try:
+            # Get list of OceanStor port details
+            ports = self.client.get_all_ports()
+            logic_type_dict = {
+                consts.PORT_LOGICTYPE_HOST:
+                    constants.PortLogicalType.SERVICE,
+                consts.PORT_LOGICTYPE_EXPANSION:
+                    constants.PortLogicalType.OTHER,
+                consts.PORT_LOGICTYPE_MANAGEMENT:
+                    constants.PortLogicalType.MANAGEMENT,
+                consts.PORT_LOGICTYPE_INTERNAL:
+                    constants.PortLogicalType.INTERNAL,
+                consts.PORT_LOGICTYPE_MAINTENANCE:
+                    constants.PortLogicalType.MAINTENANCE,
+                consts.PORT_LOGICTYPE_SERVICE:
+                    constants.PortLogicalType.SERVICE,
+                consts.PORT_LOGICTYPE_MAINTENANCE2:
+                    constants.PortLogicalType.MAINTENANCE,
+                consts.PORT_LOGICTYPE_INTERCONNECT:
+                    constants.PortLogicalType.INTERCONNECT,
+            }
+
+            port_list = []
+            for port in ports:
+                port_type = constants.PortType.OTHER
+                health_status = constants.PortHealthStatus.ABNORMAL
+                conn_status = constants.PortConnectionStatus.CONNECTED
+
+                logical_type = logic_type_dict.get(
+                    port.get('LOGICTYPE'), constants.PortLogicalType.OTHER)
+
+                if port['HEALTHSTATUS'] == consts.PORT_HEALTH_UNKNOWN:
+                    health_status = constants.PortHealthStatus.UNKNOWN
+                if port['HEALTHSTATUS'] == consts.PORT_HEALTH_NORMAL:
+                    health_status = constants.PortHealthStatus.UNKNOWN
+
+                if port['RUNNINGSTATUS'] == consts.PORT_RUNNINGSTS_UNKNOWN:
+                    conn_status = constants.PortConnectionStatus.UNKNOWN
+                if port['RUNNINGSTATUS'] == consts.PORT_RUNNINGSTS_LINKDOWN:
+                    conn_status = constants.PortConnectionStatus.DISCONNECTED
+
+                speed = port.get('RUNSPEED')        # ether -1 or M bits/sec
+                if speed != -1:
+                    speed = None
+                max_speed = port.get('MAXSPEED')
+
+                # FC
+                if port['TYPE'] == consts.PORT_TYPE_FC:
+                    port_type = constants.PortType.FC
+                    max_speed = port['MAXSUPPORTSPEED']     # in 1000 M bits/s
+
+                # FCoE
+                if port['TYPE'] == consts.PORT_TYPE_FCOE:
+                    port_type = constants.PortType.FCOE
+
+                # Ethernet
+                if port['TYPE'] == consts.PORT_TYPE_ETH:
```

Comment:
There could be a map, key/value is `consts.PORT_TYPE_ETH: constants.PortType.ETH`, so we can assign `port_type` as `port_type = new_map[consts.PORT_TYPE_PCIE]`

Reply:
Done